roachtest,workload: some adjustment to tpch tests

yuzefovich opened the pull request:
**workload: increase margin of error in tpch workload**

TPC-H spec requires decimals to have correct data up to 0.01 difference.
Previously we were checking that "as is" against the expected values.
However, the expected values themselves were calculated using
CockroachDB at some point, and there is no guarantee they are
"precisely correct". It is possible to have a data distribution such
that the expected value falls on one side of the "precise value" whereas
the actual value falls on the other side, and the difference between the
expected and the actual values will be larger than 0.01. In the worst
case, they could differ by 0.02 and still be correct, so this is what
we're doing in this commit.

Release note: None

**workload: fix tpch workload**

Previously, we were incorrectly reading the results of the query
execution - we would start reading rows and check against the expected
output. However, it is possible that there was an error on the server
while query was executing, and it can only be returned after rows.Next()
returns false via the call to rows.Err(). Now this is fixed to always
fully consume the rows and checking rows.Err(), and only in case there
was no error while executing the query, the expected check will fail the
workload.

Fixes: #41108. 

Release note: None

**roachtest: modify tpchvec test to tolerate errors**

This commit adds '--tolerate-errors' flag to tpch workload so that all
queries are run. This allows us to remove the skip list (except for the
queries 10 - has a bug, 12, and 15). The roachtest is considered failed
only if the workload gets "wrong results" error; other errors (like OOM
and unsupported features are ignored). This should reduce the amount of
noise and increase the test coverage.

This commit also reduces `workmem` from 2GiB to 1GiB to reduce the
change of OOM node crash.

Release note: None

yuzefovich committed:
workload: increase margin of error in tpch workload

TPC-H spec requires decimals to have correct data up to 0.01 difference.
Previously we were checking that "as is" against the expected values.
However, the expected values themselves were calculated using
CockroachDB at some point, and there is no guarantee they are
"precisely correct". It is possible to have a data distribution such
that the expected value falls on one side of the "precise value" whereas
the actual value falls on the other side, and the difference between the
expected and the actual values will be larger than 0.01. In the worst
case, they could differ by 0.02 and still be correct, so this is what
we're doing in this commit.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43291)
<!-- Reviewable:end -->

yuzefovich committed:
workload: fix tpch workload

Previously, we were incorrectly reading the results of the query
execution - we would start reading rows and check against the expected
output. However, it is possible that there was an error on the server
while query was executing, and it can only be returned after rows.Next()
returns false via the call to rows.Err(). Now this is fixed to always
fully consume the rows and checking rows.Err(), and only in case there
was no error while executing the query, the expected check will fail the
workload.

Release note: None

mjibson commented on pkg/workload/tpch/tpch.go:
```diff
@@ -283,9 +295,21 @@ func (w *worker) run(ctx context.Context) error {
+
+	// It is possible that we short-circuited the loop above, and we need to
+	// fully consume the rows before we might get an error from the server.
+	for rows.Next() {
```
I think a rows.Close() is a better choice here.

jordanlewis replied:
Yeah I think so too. You don't need to consume the whole thing to get its error I don't believe.

Also, hmm, is it expected from Postgres that you might have to consume the whole result set before seeing an error? I think probably not, which makes me a little bit nervous. If someone read 3 rows, decided to stop reading, but we had this weird error condition, they would have gotten incorrect results... not something to fix here but we should probably investigate this a little more.

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43291)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @mjibson)

___
*[pkg/workload/tpch/tpch.go, line 301 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43291#-LwOdnB793wiDuonwdcw-r3-301:-LwQCWBk0ZUAiKLcZ5xx:bo3vl6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea21aba9b7f34b2c97670e54e8d7e2e93f1c6c10/pkg/workload/tpch/tpch.go#L301)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I think a rows.Close() is a better choice here.
</blockquote></details>

Thanks, done.


<!-- Sent from Reviewable.io -->

madelynnblue reviewed (approved):
3rd commit lgtm

yuzefovich committed:
roachtest: modify tpchvec test to tolerate errors

This commit adds '--tolerate-errors' flag to tpch workload so that all
queries are run. This allows us to remove the skip list (except for the
queries 10 - has a bug, 12, and 15). The roachtest is considered failed
only if the workload gets "wrong results" error; other errors (like OOM
and unsupported features are ignored). This should reduce the amount of
noise and increase the test coverage.

This commit also reduces `workmem` from 2GiB to 1GiB to reduce the
change of OOM node crash.

Release note: None

yuzefovich commented:
I removed the first commit and added another one, so, for the record, Matt approved (what is currently) second commit.

asubiotto reviewed (dismissed):
> Reviewed 2 of 2 files at r4, 1 of 1 files at r5, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43291)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @mjibson, and @yuzefovich)

___
*[pkg/cmd/roachtest/tpchvec.go, line 142 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/43291#-LwSUfZw7HE_pHs4acEL:-LwSUfZw7HE_pHs4acEM:b55b2uy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2abaceccb6465f9197753a5bc97d3d5b54b2058/pkg/cmd/roachtest/tpchvec.go#L142)):*
> ```Go
> 			// ignore errors like OOM and unsupported features in order to not
> 			// short-circuit the run of this test).
> 			if strings.Contains(string(workloadOutput), "wrong result") {
> ```

It would probably be better to create an error type that is unwrapped and checked here. Checking error strings is brittle.
___
*[pkg/workload/tpch/tpch.go, line 222 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/43291#-LwSTaYp-Ci6HZ0PqUIt:-LwSTaYp-Ci6HZ0PqUIu:btkuj29) ([raw file](https://github.com/cockroachdb/cockroach/blob/75e92a280079b9cd94580549d9df74b28bdb43dc/pkg/workload/tpch/tpch.go#L222)):*
> ```Go
> 	// error (if any) and will return it below if there was no error while
> 	// executing the query.
> OUTER:
> ```

Rather than use a label here, I would probably just write it as an anonymous function:
```
consumeRowsErr := func() error {
    ...
    if err != nil {
        return err
    }
    ...
}
```
___
*[pkg/workload/tpch/tpch.go, line 298 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/43291#-LwSU7Wy09wncXilJXkC:-LwSU7Wy09wncXilJXkD:bhd4x8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/75e92a280079b9cd94580549d9df74b28bdb43dc/pkg/workload/tpch/tpch.go#L298)):*
> ```Go
> 	}
> 
> 	rows.Close()
> ```

I think `Close` returns the same error as `Err`.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43291)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @mjibson)

___
*[pkg/cmd/roachtest/tpchvec.go, line 142 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/43291#-LwSUfZw7HE_pHs4acEL:-LwU1J0X5IqFfmiYFh_A:b-go485w) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2abaceccb6465f9197753a5bc97d3d5b54b2058/pkg/cmd/roachtest/tpchvec.go#L142)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

It would probably be better to create an error type that is unwrapped and checked here. Checking error strings is brittle.
</blockquote></details>

Hm, at this point we don't have access to `error`s, we only have `[]byte`, so I think that string matching is the only way to go, no?

I did make minor modifications to the matching process.
___
*[pkg/workload/tpch/tpch.go, line 222 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/43291#-LwSTaYp-Ci6HZ0PqUIt:-LwTxHYfEfVuuM82r0nc:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/75e92a280079b9cd94580549d9df74b28bdb43dc/pkg/workload/tpch/tpch.go#L222)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Rather than use a label here, I would probably just write it as an anonymous function:
```
consumeRowsErr := func() error {
    ...
    if err != nil {
        return err
    }
    ...
}
```
</blockquote></details>

Good point, done.
___
*[pkg/workload/tpch/tpch.go, line 298 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/43291#-LwSU7Wy09wncXilJXkC:-LwTx2faA_8YeBa8pEqq:b-h24b2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/75e92a280079b9cd94580549d9df74b28bdb43dc/pkg/workload/tpch/tpch.go#L298)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think `Close` returns the same error as `Err`.
</blockquote></details>

Hm, the comments to the methods are somewhat confusing:
```
Err returns the error, if any, that was encountered during iteration. Err may be called after an explicit or implicit Close.
```
It seems to be me that `Err()` is a safer choice after calling `Close`.


<!-- Sent from Reviewable.io -->

madelynnblue reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43291)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @mjibson)

___
*[pkg/workload/tpch/tpch.go, line 298 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/43291#-LwSU7Wy09wncXilJXkC:-LwU2-NQ0lp7OvrIgi-V:b-r8sacx) ([raw file](https://github.com/cockroachdb/cockroach/blob/75e92a280079b9cd94580549d9df74b28bdb43dc/pkg/workload/tpch/tpch.go#L298)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Hm, the comments to the methods are somewhat confusing:
```
Err returns the error, if any, that was encountered during iteration. Err may be called after an explicit or implicit Close.
```
It seems to be me that `Err()` is a safer choice after calling `Close`.
</blockquote></details>

https://pkg.go.dev/database/sql?tab=doc#example-DB.QueryContext

The standard library docs have an example that calls Close then Err. I think this is the correct way.

They also do check for the err result of Close. We should probably do that too.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43291)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @mjibson)

___
*[pkg/workload/tpch/tpch.go, line 298 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/43291#-LwSU7Wy09wncXilJXkC:-LwUIywUB_-glTBaxhyL:b-l0sub5) ([raw file](https://github.com/cockroachdb/cockroach/blob/75e92a280079b9cd94580549d9df74b28bdb43dc/pkg/workload/tpch/tpch.go#L298)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

https://pkg.go.dev/database/sql?tab=doc#example-DB.QueryContext

The standard library docs have an example that calls Close then Err. I think this is the correct way.

They also do check for the err result of Close. We should probably do that too.
</blockquote></details>

Actually, just doing `rows.Close` is not sufficient - we need to call `rows.Next()` until it returns `false` (I've just verified that behavior - without fully consuming rows, `rows.Err` doesn't return the memory error I expected), so I reverted back to that.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Finally was able to run the test successfully:
```
NOTE: the comparison might not be fair because vec ON doesn't spill to disk
19:13:56 tpchvec.go:170: [q1] vec ON was faster by 126.06%: 4.95s ON vs 11.19s OFF
19:13:56 tpchvec.go:170: [q2] vec ON was faster by 144.90%: 0.98s ON vs 2.40s OFF
19:13:56 tpchvec.go:170: [q3] vec ON was faster by 72.80%: 2.50s ON vs 4.32s OFF
19:13:56 tpchvec.go:170: [q4] vec ON was faster by 148.21%: 4.19s ON vs 10.40s OFF
19:13:56 tpchvec.go:165: [q6] vec OFF was faster by 0.24%: 8.46s ON vs 8.44s OFF --- WARNING
19:13:56 tpchvec.go:170: [q7] vec ON was faster by 0.59%: 10.13s ON vs 10.19s OFF
19:13:56 tpchvec.go:170: [q8] vec ON was faster by 316.17%: 7.98s ON vs 33.21s OFF
19:13:56 tpchvec.go:170: [q9] vec ON was faster by 51.82%: 153.93s ON vs 233.69s OFF
19:13:56 tpchvec.go:170: [q11] vec ON was faster by 4.32%: 1.62s ON vs 1.69s OFF
19:13:56 tpchvec.go:170: [q13] vec ON was faster by 101.72%: 4.07s ON vs 8.21s OFF
19:13:56 tpchvec.go:170: [q14] vec ON was faster by 51.43%: 0.70s ON vs 1.06s OFF
19:13:56 tpchvec.go:170: [q16] vec ON was faster by 201.52%: 0.66s ON vs 1.99s OFF
19:13:56 tpchvec.go:170: [q17] vec ON was faster by 7.14%: 0.56s ON vs 0.60s OFF
19:13:56 tpchvec.go:170: [q18] vec ON was faster by 154.08%: 6.49s ON vs 16.49s OFF
19:13:56 tpchvec.go:170: [q19] vec ON was faster by 91.25%: 5.03s ON vs 9.62s OFF
19:13:56 tpchvec.go:170: [q20] vec ON was faster by 39.27%: 9.37s ON vs 13.05s OFF
19:13:56 tpchvec.go:170: [q21] vec ON was faster by 99.26%: 13.52s ON vs 26.94s OFF
19:13:56 tpchvec.go:170: [q22] vec ON was faster by 32.58%: 0.89s ON vs 1.18s OFF
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43291#-:-LwUcpL45dPbWSnTlQSP:bnv5yvg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @mjibson)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Friendly ping @jordanlewis, I'm not sure if you'll have time, but it'd be great to make `tpchvec` roachtest useful again. I hope the review shouldn't take too long (Matt has already approved the second commit).

jordanlewis commented on pkg/workload/tpch/tpch.go:
```diff
@@ -217,74 +231,105 @@ func (w *worker) run(ctx context.Context) error {
+									// occur:
+									//   expected < "ideal" < actual
+									//   "ideal" - expected < 0.01 && actual - "ideal" < 0.01
+									// so in the worst case, actual and expected might differ by
```
This seems sketchy - when would ideal not exactly match expected?

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43291)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @mjibson)

___
*[pkg/workload/tpch/tpch.go, line 301 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43291#-LwOdnB793wiDuonwdcw-r3-301:-Lx21UBe6eFZPFUuRL8k:b13txmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea21aba9b7f34b2c97670e54e8d7e2e93f1c6c10/pkg/workload/tpch/tpch.go#L301)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Yeah I think so too. You don't need to consume the whole thing to get its error I don't believe.

Also, hmm, is it expected from Postgres that you might have to consume the whole result set before seeing an error? I think probably not, which makes me a little bit nervous. If someone read 3 rows, decided to stop reading, but we had this weird error condition, they would have gotten incorrect results... not something to fix here but we should probably investigate this a little more.
</blockquote></details>

I agree that this is unexpected expectations from the user. I'll try to debug why fully consuming `rows` seemed necessary.
___
*[pkg/workload/tpch/tpch.go, line 292 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/43291#-LwOdnB793wiDuonwdcw-r9-292:-Lx1zPT645RxoZ2CCU9k:bxe4oh3) ([raw file](https://github.com/cockroachdb/cockroach/blob/80e2c6ddb4b3c11ab2e290a337b1b04a7f5f5412/pkg/workload/tpch/tpch.go#L292)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This seems sketchy - when would ideal not exactly match expected?
</blockquote></details>

Different data distributions cause different rounding errors when aggregating, that's why on several runs of the test we are expected to get slightly different results. According to the spec, those results should be within 0.01 of the "exact ideal" output. Now, I don't know what that "exact ideal" output is, and - if I remember correctly - when creating `expected_rows.go` file that contains the "expected" output for the purposes of the verification, I took the results of running the queries against CRDB (and roughly confirmed them with https://hyper-db.de/interface.html).

I agree that such verification will not fail some cases when it should (for example, "exact ideal" is 0.01, "expected" is 0.005, actual output is 0.021 - in this case our output differs from the "exact ideal" by more than 0.01 yet it is within 0.02 of the "expected"). However, I think this is an extremely small part of the solution space, and overall we will get significant win from having less noise in this test.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTRs!

bors r+
